src/ploneintranet/theme/utils.py

```diff
@@ -0,0 +1,116 @@
+from zope import component
+from zope.schema import getFieldNames
+from zope.schema import interfaces
+
+import logging
+
+log = logging.getLogger(__name__)
+
+
+def get_dexterity_schemas(context=None, portal_type=None):
+    """ Utility method to get all schemas for a dexterity object.
+
+        IMPORTANT: Either context or portal_type must be passed in, NOT BOTH.
+        The idea is that for edit forms context is passed in and for add forms
+        where we don't have a valid context we pass in portal_type.
+
+        This builds on getAdditionalSchemata, which works the same way.
+    """
+    if context is not None:
+        portal_type = context.portal_type
+
+    fti = component.getUtility(IDexterityFTI, name=portal_type)
+    schemas = [fti.lookupSchema()]
+    for behavior_schema in \
+            getAdditionalSchemata(context=context,
+                                  portal_type=portal_type):
+        if behavior_schema is not None:
+            schemas.append(behavior_schema)
+    return schemas
+
+
+def dexterity_update(obj, request=None):
+    """
+    Utility method to update the fields of all the schemas of the Dexterity
+    object 'obj'.
+    """
+    modified = False
+    if not request:
+        request = obj.REQUEST
+    # Call processInputs to decode strings to unicode, otherwise the
+    # z3c.form dataconverters complain.
+    processInputs(request)
+    errors = []
+    for schema in get_dexterity_schemas(context=obj):
+        for name in getFieldNames(schema):
+            field = schema[name]
+            autoform_widgets = schema.queryTaggedValue(WIDGETS_KEY, default={})
+            if name in autoform_widgets:
+                widgetclass = resolveDottedName(autoform_widgets[name])
+                widget = widgetclass(field, request)
+            else:
+                widget = component.getMultiAdapter(
+                    (field, request), IFieldWidget)
+
+            widget.context = obj
+            value = field.missing_value
+            widget.update()
+            try:
+                raw = widget.extract()
+            except MultipleErrors, e:
+                errors.append(e)
+                log.warn("Multiple errors while extracting field: %s" % name)
+                continue
+
+            if raw is NOT_CHANGED:
+                continue
+
+            if raw is NO_VALUE:
+                continue
+
+            value = IDataConverter(widget).toFieldValue(safe_unicode(raw))
+
+            try:
+                field.validate(value)
+            except interfaces.RequiredMissing, e:
+                errors.append(e)
+                log.warn("Required field have missing value: %s" % name)
+                # XXX: we might not want to continue here, since we then remove
+                # the ability to remove field values. Client side validation
+                # should prevent this situation from arising, but it's not yet
+                # perfect.
+                # continue
```

Comment:
@pilz: here it say we "might not want to continue", but continue is commented out anyway. What is the rationale here?